Drop token exchange provider implementation in favor of token-exchange-standard:v2.

chenkins opened the pull request:
Closes #66 

Client implementation: https://github.com/shift7-ch/katta-clientlib/pull/123

```mermaid
sequenceDiagram
    participant User
    participant Katta Client
    participant Keycloak
    participant STS
    participant S3
    Katta Client ->> Keycloak: (1) /authorize
    Keycloak -->> Katta Client: (2) redirect
    Katta Client --) User: (2) prompt in browser
    User --) Keycloak: (3) User/Password
    Keycloak --) Katta Client: (4) Authorization Code
    Katta Client ->> Keycloak: (5): /token
    Note over Keycloak: client_id: cryptomator
    Keycloak -->> Katta Client: (6) access_token: OIDC Access Token (JWT)
    Note over Katta Client: { aud: ["cryptomator", "cryptomatorvaults"], <br/>azp: "cryptomator, ... }
    Katta Client --> Keycloak: (7) /token
    Note over Keycloak: grant_type: token-exchange<br/>client_id: cryptomatorvaults <br/> subject_token: <access_token><br/> scope: <vaultId>
    Keycloak -->> Katta Client: (8) exchanged_access_token: OIDC Access Token (JWT)
    alt MinIO STS
        Note over Katta Client: { aud: "cryptomatorvaults", client_id: <vaultId> }
        Katta Client ->> STS: (9) AssumeRoleWithWebIdentity(exchanged_access_token)
        Note over STS: IdP: client_id=cryptomatorvaults -> policy: ${jwt:client_id}
        STS -->> Katta Client: (10) AccessKeyId, SecretKey
        Note over Katta Client: { Action: s3:PutObject, ..., Resource: "arn:aws:s3:::katta<vaultId>/*"}
    else AWS STS
        Note over Katta Client: { "aud": "cryptomatorvaults", "https://aws.amazon.com/tags": {"principal_tags":{"<vaultId>":[""]},"TransitiveTagKeys":["<vaultId>"]}, ...}
        Katta Client ->> MinIO STS: (9) AssumeRoleWithWebIdentity(exchanged_access_token)
        Note over STS: 
        MinIO STS -->> Katta Client: (10) AccessKeyId, SecretKey
        Note over Katta Client: { Action: [sts:AssumeRole, sts:TagSession], Resource: "arn:aws:iam::...:role/katta_chain_02"}
        Katta Client ->> MinIO STS: (11) AssumeRole(AccessKeyId, SecretKey, roleArn="arn:aws:iam::...:role/katta_chain_02", tag.name=VaultRequested, tag.value=<vaultId>)
        Note over STS: "Condition": { "ForAnyValue:StringEquals": { "sts:TransitiveTagKeys": "${aws:RequestTag/VaultRequested}" } }
        MinIO STS -->> Katta Client: (12) AccessKeyId, SecretKey
        Note over Katta Client: { Action: s3:PutObject, ..., Resource: "arn:aws:s3:::katta<vaultId>/*"}
    end
    Katta Client ->> S3: (13) /list-bucket

```

References:
* <a href="https://www.keycloak.org/securing-apps/token-exchange#_standard-token-exchange">Standard token exchange: version 2 (V2)</a>
* https://min.io/docs/minio/linux/administration/identity-access-management/oidc-access-management.html#oidc-policy-variables

chenkins committed:
Drop token exchange provider implementation in favor of token-exchange-standard:v2.

chenkins commented on keycloak/src/test/java/cloud/katta/KattaDevRealmIT.java:
```diff
@@ -31,7 +31,7 @@ public class KattaDevRealmIT {
 	@Test
 	@Disabled
```
enable, merge with testKattaTokenExchange?

chenkins commented on backend/src/test/java/org/cryptomator/hub/cipherduck/KeycloakContainerResource.java:
```diff
@@ -12,7 +12,7 @@ public class KeycloakContainerResource implements QuarkusTestResourceLifecycleMa
 	@Override
 	public Map<String, String> start() {
-		container = new KeycloakContainer("quay.io/keycloak/keycloak:26.1.5")
+		container = new KeycloakContainer("quay.io/keycloak/keycloak:26.2.2")
```
extract version?

chenkins commented on keycloak/src/test/java/cloud/katta/KattaTokenExchangeIT.java:
```diff
@@ -158,85 +195,71 @@ public void testCipherduckTokenExchange(final boolean spiEnabled, final boolean
 				final String exchangedAccessToken = given()
-						.formParam("client_id", "cryptomator")
-						.formParam("audience", "cryptomatorvaults")
+						.formParam("client_id", "cryptomatorvaults")
+						.formParam("client_secret", "")
 						.formParam("grant_type", "urn:ietf:params:oauth:grant-type:token-exchange")
+						.formParam("subject_token_type", "urn:ietf:params:oauth:token-type:access_token")
```
update clientlib

chenkins commented on keycloak/README.md:
Updaate top-level README

chenkins committed:
Update README, dropping token exchange provider implementation in favor of token-exchange-standard:v2.

chenkins committed:
Bump testcontainers-keycloak to 3.7.0.

chenkins committed:
Cleanup.

chenkins committed:
Code cleanup.

chenkins committed:
Re-enable/fix dev realm integration test.

chenkins committed:
Only cryptomatorvaults needs standard.token.exchange.enabled.

chenkins committed:
Cleanup.

chenkins committed:
Test claims from added protocol mappers.

chenkins committed:
Documentation standard token exchange v2.

chenkins committed:
Cleanup protocol mapper naming.

chenkins committed:
Add test illustrating serviceAccountsEnabled. Disable all grants except for token exchange for cryptomatorvaults client.

chenkins committed:
Merge tests and cleanup documentation.

overheadhunter reviewed (approved):
Looks good, so glad this can now be done with an unpatched Keycloak.

The sequence diagram is really helpful and needs to be on a world heritage list, however the STS and Minio STS got mixed up 😉

ylangisc commented on backend/pom.xml:
```diff
@@ -286,7 +286,7 @@
               <target>
-                <!-- undo to dev.json because of https://github.com/keycloak/keycloak/issues/37569 for Keycloak 26.1.5, remove when fixed -->
+                <!-- rename to dev.json because of https://github.com/keycloak/keycloak/issues/37569 -->
                 <move file="${project.build.directory}/classes/dev-realm.json"
```
Referenced issue is closed as import seems to be by filename convention. Create subfolder `realms` and leave the filename `dev.json` to get rid of this workaround?

chenkins replied:
Upstream uses `dev-realm.json`.
When started via `mvn clean quarkus:dev`, the file name does not cause a problem (probably, there is some pre-processing there).
When started via docker compose with volume mapping, then the expected behaviour seems indeed to be to fail if `XYZ-realm.json` does not contain the realm `XYZ`, see https://github.com/keycloak/keycloak/issues/37569#issuecomment-2700173095. 

So the change keeps the diff to upstream minimal.

ylangisc commented on keycloak/pom.xml:
```diff
@@ -14,8 +14,11 @@
         <failsafe-plugin.version>3.2.5</failsafe-plugin.version>
 
-        <keycloak.version>26.1.5</keycloak.version>
+        <keycloak.version>26.2.2</keycloak.version>
```
Update to latest version 26.2.5?

chenkins replied:
Keeping in sync with our upstream branch https://github.com/cryptomator/hub/blob/feature/uvf/backend/src/main/resources/application.properties#L36 for now. develop has 26.2.25 already: https://github.com/cryptomator/hub/blob/develop/backend/src/main/resources/application.properties#L36. Will update as soon as pr is merged or develop is merged into fb.

overheadhunter reviewed (changes requested):
For the sake of documenting results of my research after our discussion today:

Since the token exchange MUST be done by a confidential client, it should NOT be done by our desktop client.

> The entity that makes the request to exchange tokens is considered the client in the context of the token exchange interaction. However, that does not restrict usage of this profile to traditional OAuth clients. An OAuth resource server, for example, might assume the role of the client during token exchange in order to trade an access token that it received in a protected resource request for a new token that is appropriate to include in a call to a backend service.
> (https://datatracker.ietf.org/doc/html/rfc8693#section-1-5)

So the _OAuth Resource Server_ (i.e. Katta server) can receive the access token by the _OAuth Client_ (i.e. Katta client) in order to do a token exchange with the _OAuth Authorization Server_ (Keycloak).

There is an example for this in [Chapter 2.3](https://datatracker.ietf.org/doc/html/rfc8693#section-2.3):

> [...] the resource server receiving a protected resource request containing an OAuth access token in the Authorization header
> [...]
> The value of the `subject_token` parameter carries the access token, and the value of the `subject_token_type` parameter indicates that it is an OAuth 2.0 access token. The resource server, acting in the role of the client, uses its identifier and secret to authenticate to the authorization server using the HTTP Basic authentication scheme.

I adjusted Steps 7+8:

```mermaid
sequenceDiagram
    participant User
    participant Katta Client
    participant Katta Server
    participant Keycloak
    participant STS
    participant S3
    Katta Client ->> Keycloak: (1) /authorize
    Keycloak -->> Katta Client: (2) redirect
    Katta Client --) User: (2) prompt in browser
    User --) Keycloak: (3) User/Password
    Keycloak --) Katta Client: (4) Authorization Code

    Katta Client ->> Keycloak: (5): /token
    Note over Keycloak: client_id: cryptomator
    Keycloak -->> Katta Client: (6) access_token: OIDC Access Token (JWT)

    Note over Katta Client: { aud: ["cryptomator"], <br/>azp: "cryptomator, ... }
    

    Katta Client ->> Katta Server: (7a) /token
    Katta Server ->> Keycloak: (7b): /token    
    Note over Keycloak: grant_type: token-exchange<br/>client_id: cryptomatorvaults <br/> subject_token: <access_token><br/> scope: <vaultId>
    Keycloak -->> Katta Server: (8a) exchanged_access_token: OIDC Access Token (JWT)
    Katta Server ->> Katta Client: (8b) exchanged_access_token: OIDC Access Token (JWT)
    alt MinIO STS
        Note over Katta Client: { aud: "cryptomatorvaults", client_id: <vaultId> }
        Katta Client ->> MinIO STS: (9) AssumeRoleWithWebIdentity(exchanged_access_token)
        Note over MinIO STS: IdP: client_id=cryptomatorvaults -> policy: ${jwt:client_id}
        MinIO STS -->> Katta Client: (10) AccessKeyId, SecretKey
        Note over Katta Client: { Action: s3:PutObject, ..., Resource: "arn:aws:s3:::katta<vaultId>/*"}
    else AWS STS
        Note over Katta Client: { "aud": "cryptomatorvaults", "https://aws.amazon.com/tags": {"principal_tags":{"<vaultId>":[""]},"TransitiveTagKeys":["<vaultId>"]}, ...}
        Katta Client ->> STS: (9) AssumeRoleWithWebIdentity(exchanged_access_token)
        STS -->> Katta Client: (10) AccessKeyId, SecretKey
        Note over Katta Client: { Action: [sts:AssumeRole, sts:TagSession], Resource: "arn:aws:iam::...:role/katta_chain_02"}
        Katta Client ->> STS: (11) AssumeRole(AccessKeyId, SecretKey, roleArn="arn:aws:iam::...:role/katta_chain_02", tag.name=VaultRequested, tag.value=<vaultId>)
        Note over STS: "Condition": { "ForAnyValue:StringEquals": { "sts:TransitiveTagKeys": "${aws:RequestTag/VaultRequested}" } }
        STS -->> Katta Client: (12) AccessKeyId, SecretKey
        Note over Katta Client: { Action: s3:PutObject, ..., Resource: "arn:aws:s3:::katta<vaultId>/*"}
    end
    Katta Client ->> S3: (13) /list-bucket
```

chenkins commented on backend/src/main/resources/dev-realm.json:
```diff
@@ -466,87 +325,60 @@
     },
     {
-      "id": "cryptomatorvaults",
       "clientId": "cryptomatorvaults",
```
What's the lifetime of the exchanged token? It can/should be much shorter than the lifetime of the initial token: https://datatracker.ietf.org/doc/html/rfc8693#section-2.1

overheadhunter committed:
add `POST /api/storage/s3-token`

to exchange token issued for `cryptomator` public client for `cryptomatorvaults` client

overheadhunter committed:
token exchange officially supported since KC 26.2

see https://www.keycloak.org/2025/04/keycloak-2620-released

overheadhunter committed:
token exchange officially supported since KC 26.2

see https://www.keycloak.org/2025/04/keycloak-2620-released

overheadhunter committed:
token exchange officially supported since KC 26.2

see https://www.keycloak.org/2025/04/keycloak-2620-released

overheadhunter commented:
#77 implements the changes [proposed above](#pullrequestreview-2900666291)

overheadhunter committed:
disable needless S3 dev services

because localstack container requires rootful container execution,
which is disabled by default in podman, unless activated via `podman machine set --rootful=true`

also speeding up builds a bit

chenkins committed:
Add test showing that Keycloak does not down-scope access token upon token refresh.

overheadhunter committed:
Revert aa5925e, 2309ba3, df400cf

overheadhunter committed:
`dev-realm.json` → `cryptomator-realm.json`

chenkins commented on keycloak/src/test/java/cloud/katta/KattaTokenExchangeIT.java:
```diff
@@ -0,0 +1,681 @@
+ *  <li>(P1) <code>cryptomator</code> client is public, <code>cryptomatorvaults</code> is confidential and has standard token exchange enabled.</li>
+ * 	<li>(P2) the initial token contains audiences <code>cryptomator</code> and <code>cryptomatorvaults</code> and <code>azp</code>=<code>cryptomator</code></li>
+ * 	<li>(P3) the initial token contains realm roles</li>
+ * 	<li>(P4) the exchanged token contains audience <code>cryptomatorvaults</code> and <code>azp</code>=<code>cryptomatorvaults</code> </li>
```
Add reference why `azp=cryptomatorvaults` is necessary.

overheadhunter committed:
exchange token as `cryptomatorvaults`

making sure `azp` header is `cryptomatorvaults`

```
{
  "exp": 1752145041,
  "iat": 1752144741,
  "auth_time": 1752144741,
  "jti": "ntrtte:5f48bed6-906e-4078-a7d1-63ff6ebef7da",
  "iss": "http://localhost:37965/realms/cryptomator",
  "sub": "758babab-ec50-4d0d-ae2d-828625e54fba",
  "typ": "Bearer",
  "azp": "cryptomatorvaults",
  "sid": "8f12aa49-f138-4664-9d2e-1ba3a520b30e",
  "scope": "address",
  "address": {}
}
```

overheadhunter committed:
exchange token as `cryptomatorvaults`

making sure `azp` header is `cryptomatorvaults`

```
{
  "exp": 1752145041,
  "iat": 1752144741,
  "auth_time": 1752144741,
  "jti": "ntrtte:5f48bed6-906e-4078-a7d1-63ff6ebef7da",
  "iss": "http://localhost:37965/realms/cryptomator",
  "sub": "758babab-ec50-4d0d-ae2d-828625e54fba",
  "typ": "Bearer",
  "azp": "cryptomatorvaults",
  "sid": "8f12aa49-f138-4664-9d2e-1ba3a520b30e",
  "scope": "address",
  "address": {}
}
```

overheadhunter committed:
parameterized: `/api/storage/s3-token?vault=foo`

overheadhunter committed:
adjust credentials

overheadhunter committed:
re-add `"aud": "cryptomatorvaults"`:

```
{
  "exp": 1752147102,
  "iat": 1752146802,
  "auth_time": 1752146802,
  "jti": "ntrtte:2377d26d-12e7-4f97-9afb-7bcfc23313f3",
  "iss": "http://localhost:34455/realms/cryptomator",
  "aud": "cryptomatorvaults",
  "sub": "e3545ec6-4421-4ca6-858a-3b77000043cc",
  "typ": "Bearer",
  "azp": "cryptomatorvaults",
  "sid": "812a1d9b-342a-4cfb-b8b3-d3c5824bc08f",
  "scope": "address",
  "address": {}
}
```

overheadhunter committed:
adjust credentials

overheadhunter committed:
reordered lines to minimize diff

chenkins committed:
Merge pull request #77 from shift7-ch/feature/proxies-token-exchange

add token exchange proxy via `POST /api/storage/s3-token`

chenkins committed:
Fix realm import in test.

chenkins committed:
Merge branch 'feature/cipherduck-uvf' of github.com:shift7-ch/katta-server into feature/66-use-token-exchange-standard-v2

chenkins committed:
Add return type for s3-token endpoint.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/api/cipherduck/KeycloakTokenExchangeApi.java:
```diff
@@ -0,0 +1,26 @@
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+import org.keycloak.representations.AccessTokenResponse;
```
Are you sure this is intended to be used as public API?

chenkins committed:
Use custom return type for s3-token endpoint.
